Adds DepthSensorToLcmPointCloudMessage.

liangfok opened the pull request:
Resolves #1586.

Here's a visualization of the first octant point cloud in `drake-visualizer` when the point size is manually set to 5 and colored black:

# First Octant Spec Visualization

```
$ bazel run //drake/systems/sensors:depth_sensor_to_lcm_point_cloud_message_demo
```

<img width="1549" alt="screen shot 2017-04-09 at 9 53 09 pm" src="https://cloud.githubusercontent.com/assets/1388098/24843252/0bc01bc6-1d6f-11e7-92f7-5c6b1e3416aa.png">

# XY Planar Spec Visualization

```
$ bazel run //drake/systems/sensors:depth_sensor_to_lcm_point_cloud_message_demo -- --spec=xy_planar
```

<img width="1549" alt="screen shot 2017-04-10 at 10 04 32 am" src="https://cloud.githubusercontent.com/assets/1388098/24865426/3611a44c-1dd5-11e7-8155-7a22c8a90206.png">

# XZ Planar Spec Visualization

```
$ bazel run //drake/systems/sensors:depth_sensor_to_lcm_point_cloud_message_demo -- --spec=xz_planar
```

<img width="1549" alt="screen shot 2017-04-10 at 10 07 43 am" src="https://cloud.githubusercontent.com/assets/1388098/24865561/9e1bf2ae-1dd5-11e7-99dd-cf4b958d5a53.png">

# XYZ Spherical Spec Visualization

```
$ bazel run //drake/systems/sensors:depth_sensor_to_lcm_point_cloud_message_demo -- --spec=xyz_spherical
```

<img width="1549" alt="screen shot 2017-04-10 at 10 10 38 am" src="https://cloud.githubusercontent.com/assets/1388098/24865687/feacd278-1dd5-11e7-8db3-fc2137229cac.png">

# X Linear Spec Visualization

```
bazel run //drake/systems/sensors:depth_sensor_to_lcm_point_cloud_message_demo -- --spec=x_linear
```

<img width="1549" alt="screen shot 2017-04-10 at 10 14 06 am" src="https://cloud.githubusercontent.com/assets/1388098/24865848/7ca5a9d4-1dd6-11e7-8b7f-cdb12b1d3711.png">

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5779)
<!-- Reviewable:end -->

liangfok commented:
+@kunimatsu-tri for feature review please. Thanks!

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-:-KhNR_LG_tkrG7g1k-B5:bk1y2no)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Sure! I will work on this as soon as OKR review session finishes.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-:-KhNyoo74Kby9UIdtNQI:boej9ii)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Many thanks!

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-:-KhO0FhGrNEqv-hohMgT:b-adxy90)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Finished the first round.

---

Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/sensors/depth_sensor_vis.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOx9d2T25z_89DXJox:-KhOx9d3L44Z7i7nkzKx:bvl9fpg) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/depth_sensor_vis.h#L22)):*
> ```objc
> /// provided to the constructor.
> ///
> class DepthSensorVis : public systems::LeafSystem<double> {
> ```

Have you considered using `LcmPublisherSystem` instead of writing brand new class from scratch?:
https://github.com/RobotLocomotion/drake/blob/master/drake/systems/lcm/lcm_publisher_system.h

If so, I'm curious to know what made you choose this design since I'm thinking to do the similar thing with `RgbdCamera`.

---

*[drake/systems/sensors/depth_sensor_vis.h, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOtGLU71ZcS1DcWXR4:-KhOtGLU71ZcS1DcWXR5:bzebp73) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/depth_sensor_vis.h#L36)):*
> ```objc
>   /// @param lcm The LCM interface for publishing LCM messages. The lifespan of
>   /// this parameter must exceed that of this object.
>   DepthSensorVis(const std::string& name, const DepthSensorSpecification& spec,
> ```

'name` sounds a little bit ambiguous to me. How about choosing clearer name e.g. `sensor_name`? (as mentioned in the class description)

---

*[drake/systems/sensors/depth_sensor_vis.h, line 41 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOu7PzPjRD_1LW5H-5:-KhOu7Q-eZS3EkZXKxz9:b-ulxmqr) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/depth_sensor_vis.h#L41)):*
> ```objc
>   /// Returns a descriptor of the input port containing a DepthSensorOutput.
>   const InputPortDescriptor<double>& depth_readings_input_port()
>       const;
> ```

BTW, this seems unnecessary line break. How about placing this on the line above?

---

*[drake/systems/sensors/test/depth_sensor_vis_app.cc, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOuzddzJ4zHl22OHZ5:-KhOuzde1d7t_KBA6DYm:b-3ixgg5) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/test/depth_sensor_vis_app.cc#L21)):*
> ```cpp
> namespace sensors {
> 
> int main(int argc, char* argv[]) {
> ```

If I understand correctly, this isn't an unit test but an example code.
If so, shouldn't this be in `example/(somewhere)` directory instead of `test` directory?

---

*[drake/systems/sensors/test/depth_sensor_vis_test.cc, line 8 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOz2-cCbueHiJPWMgx:-KhOz2-dqtfYN-feBlQc:b-jceh7x) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/test/depth_sensor_vis_test.cc#L8)):*
> ```cpp
> 
> #include <gtest/gtest.h>
> ```

I believe both `gtest/gtest.h` and `bot_core/pointcloud_t.hpp` should be regarded as "4. Other libraries' .h files.": 
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

If this is true, this line break must be removed and two header files has to be sorted alphabetically.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-:-KhP0SgrWZV7gGKtCt4C:b-c9otiw)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/depth_sensor_vis.h, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOtGLU71ZcS1DcWXR4:-KhP0emGYfSM9UBzwB0V:bjt3574) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/depth_sensor_vis.h#L36)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

'name` sounds a little bit ambiguous to me. How about choosing clearer name e.g. `sensor_name`? (as mentioned in the class description)
</blockquote></details>

Oops, sorry. 

`name` sounds a little bit ambiguous to me. How about choosing clearer name e.g. `sensor_name`? (as mentioned in the class description)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@kunimatsu-tri I addressed your comments. PTAL. Thanks!

---

Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/sensors/depth_sensor_vis.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOx9d2T25z_89DXJox:-KhP35TqAPjvmsQzCBab:b-nye48y) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/depth_sensor_vis.h#L22)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Have you considered using `LcmPublisherSystem` instead of writing brand new class from scratch?:
https://github.com/RobotLocomotion/drake/blob/master/drake/systems/lcm/lcm_publisher_system.h

If so, I'm curious to know what made you choose this design since I'm thinking to do the similar thing with `RgbdCamera`.
</blockquote></details>

Good question. Upon further reflection, I now believe using `LcmPublisherSystem` is preferable since it has additional features like support for a publishing period. I updated this system to use it.

---

*[drake/systems/sensors/depth_sensor_vis.h, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOtGLU71ZcS1DcWXR4:-KhP7e-rtIqaCjsLtWSa:b-fzj022) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/depth_sensor_vis.h#L36)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Oops, sorry. 

`name` sounds a little bit ambiguous to me. How about choosing clearer name e.g. `sensor_name`? (as mentioned in the class description)

</blockquote></details>

Done (removed).

---

*[drake/systems/sensors/depth_sensor_vis.h, line 41 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOu7PzPjRD_1LW5H-5:-KhP7iL_6e4wEDjpUIop:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/depth_sensor_vis.h#L41)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

BTW, this seems unnecessary line break. How about placing this on the line above?
</blockquote></details>

Done.

---

*[drake/systems/sensors/test/depth_sensor_vis_test.cc, line 8 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOz2-cCbueHiJPWMgx:-KhP7osOmfrW-ObjBPa4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/test/depth_sensor_vis_test.cc#L8)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

I believe both `gtest/gtest.h` and `bot_core/pointcloud_t.hpp` should be regarded as "4. Other libraries' .h files.": 
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

If this is true, this line break must be removed and two header files has to be sorted alphabetically.
</blockquote></details>

Done.

---

*[drake/systems/sensors/test/depth_sensor_vis_app.cc, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOuzddzJ4zHl22OHZ5:-KhP8GD_LqlSqYc1saKl:b-iq6bx7) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/test/depth_sensor_vis_app.cc#L21)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

If I understand correctly, this isn't an unit test but an example code.
If so, shouldn't this be in `example/(somewhere)` directory instead of `test` directory?
</blockquote></details>

Done. I moved it to be in the same directory as `depth_sensor_vis` in an effort to keep closely related files together.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-:-KhScRp1F9aTcyKujKa1:bslvl2g)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Finished the second round.

---

Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/sensors/depth_sensor_vis.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOx9d2T25z_89DXJox:-KhShkYQBCk3OnuLNNQ8:b8b16ge) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/depth_sensor_vis.h#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Good question. Upon further reflection, I now believe using `LcmPublisherSystem` is preferable since it has additional features like support for a publishing period. I updated this system to use it.
</blockquote></details>

I'm glad that my question gave you a clue for improvement and to see that the code simplified very much.

Then, does it make sense to change the class name `DepthSensorVis` to exactly what this class is doing?
For instance, something like `DepthSensorOutputLcmMessageConverter` (I know this is too long though).

Since I believe this class will have much wider usage than `Vis`.

---

*[drake/systems/sensors/test/depth_sensor_vis_test.cc, line 63 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhSmMc1wCmxw7F1yij0:-KhSmMc2qqUrjwqUFzP-:b-75f6gb) ([raw file](https://github.com/robotlocomotion/drake/blob/70181e597c9e0495284d64e7cd407e1d2e775c96/drake/systems/sensors/test/depth_sensor_vis_test.cc#L63)):*
> ```cpp
> const int kZ(2);
> 
> TEST_F(TestDepthSensorVis, Octant1Test) {
> ```

This took a bit for me to understand what's done here at a glance. 
Can you add a comment for each `TEST_F` to briefly describe what each test is doing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-:-KhSnQyhVo1r5F_ZOW4v:bkha11w)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@kunimatsu-tri I addressed your comments. PTAL. Thanks!

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/sensors/depth_sensor_vis.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOx9d2T25z_89DXJox:-KhTDIYGVvi-cbxw4-FO:b4hz5k1) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/depth_sensor_vis.h#L22)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

I'm glad that my question gave you a clue for improvement and to see that the code simplified very much.

Then, does it make sense to change the class name `DepthSensorVis` to exactly what this class is doing?
For instance, something like `DepthSensorOutputLcmMessageConverter` (I know this is too long though).

Since I believe this class will have much wider usage than `Vis`.

</blockquote></details>

I renamed it to be `DepthSensorToPointCloudMessage`.

---

*[drake/systems/sensors/test/depth_sensor_vis_test.cc, line 63 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhSmMc1wCmxw7F1yij0:-KhTFofIUQV1siQ9pei5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/70181e597c9e0495284d64e7cd407e1d2e775c96/drake/systems/sensors/test/depth_sensor_vis_test.cc#L63)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

This took a bit for me to understand what's done here at a glance. 
Can you add a comment for each `TEST_F` to briefly describe what each test is doing?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-:-KhTFrmdvEA9pIF-XXOE:bslvl2g)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
I've left one comment, but <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for the rest.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-:-KhTOjGXGYaFdQWVxsND:bevww19)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/depth_sensor_to_point_cloud_message.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOx9d2T25z_89DXJox:-KhTI90RDf3AArdP6LHY:be6u1jj) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/depth_sensor_to_point_cloud_message.h#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I renamed it to be `DepthSensorToPointCloudMessage`.
</blockquote></details>

BTW, have you consider explicitly having `Lcm` in the class name? I'm asking because just `Message` has wide meaning and someone may make `DepthSensorToPointCloudRosMessage` in the future, then isn't the current name ambiguous?  If you're sure that will not happen, I'm fine with this name.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@kunimatsu-tri, thanks for the reviews!

+@sherm1 for platform review please.

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-KhOx9d2T25z_89DXJox:-KhTX5qG6b183T0VTuY1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/78ed2660345263f3299304fe455b997dc796aaa3/drake/systems/sensors/depth_sensor_to_lcm_point_cloud_message.h#L22)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

BTW, have you consider explicitly having `Lcm` in the class name? I'm asking because just `Message` has wide meaning and someone may make `DepthSensorToPointCloudRosMessage` in the future, then isn't the current name ambiguous?  If you're sure that will not happen, I'm fine with this name.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-:-KhTX8D--5NN3upuRYnB:b-k88v7o)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@drake-jenkins-bot retest this please

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> except for a few mysterious CI failures.

---

Reviewed 1 of 16 files at r2, 4 of 15 files at r3, 11 of 11 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-:-KhTr8f1MdgZWYLFO4-D:b-wg4qeh)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
BTW, this is really cool @liangfok!

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5779#-:-KhTsrC2fO2Ma0ilGZ-j:b-4w996)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@drake-jenkins-bot  linux-gcc-bazel-experimental please.

liangfok commented:
Thanks @sherm1 ! 

@drake-jenkins-bot  linux-xenial-gcc-ninja-experimental please.
